server/pulp/server/controllers/repository.py

```diff
@@ -186,20 +194,39 @@ def has_all_units_downloaded(repository):
     :return: True if completely downloaded
     :rtype: bool
     """
-    types = {}
-    q_set = model.RepositoryContentUnit.objects(repo_id=repository.repo_id)
-    for association in q_set.only('unit_id', 'unit_type_id'):
-        unit_ids = types.setdefault(association.unit_type_id, [])
-        unit_ids.append(association.unit_id)
-    for type_id, unit_ids in types.items():
-        _model = plugin_api.get_unit_model_by_id(type_id)
-        for page in paginate(unit_ids):
-            q_set = _model.objects(id__in=page, downloaded=False)
-            if q_set.count():
-                return False
+    for qs in _get_mongoengine_repo_querysets(str(repository.repo_id)):
+        if qs(downloaded=False).count():
+            return False
     return True
 
 
+def _get_mongoengine_repo_querysets(repo_id, repo_content_unit_q=None):
```

Comment:
This function does two things. It figures out what modeled types are in a repo, and then it uses those models to get query sets from another function. Both sound broadly useful, and not necessarily tied to each other, so I suggest breaking this up into two functions. Something like:

```
def get_model_classes(repo_id):
    unit_types = model.RepositoryContentUnit.objects(repo_id=repo_id).distinct('unit_type_id')
    unit_models = [plugin_api.get_unit_model_by_id(type_id) for type_id in unit_types]
    # Filter any non-MongoEngine content types.
    return [m for m in unit_models if m is not None]


def get_unit_model_query_sets_across_types(repo_id, repo_content_unit_q):
    """Maybe you can find a less-wordy name that matches the other names"""
    for unit_model in get_model_classes(repo_id):
        query_sets = get_unit_model_querysets(repo_id, unit_model, repo_content_unit_q)
        for query_set in query_sets:
            yield query_set
```